File: skellig-test-step-processing-tcp/src/main/java/org/skellig/teststep/processor/tcp/TcpTestStepProcessor.java
package org.skellig.teststep.processor.tcp;

import com.typesafe.config.Config;
import org.skellig.teststep.processing.converter.TestStepResultConverter;
import org.skellig.teststep.processing.processor.BaseTestStepProcessor;
import org.skellig.teststep.processing.processor.TestStepProcessor;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.validation.TestStepResultValidator;
import org.skellig.teststep.processor.tcp.model.TcpDetails;
import org.skellig.teststep.processor.tcp.model.TcpTestStep;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static org.skellig.task.TaskUtils.runTask;

public class TcpTestStepProcessor extends BaseTestStepProcessor<TcpTestStep> {

    private Map<String, TcpChannel> tcpChannels;

    private TcpTestStepProcessor(Map<String, TcpChannel> tcpChannels,
                                 TestScenarioState testScenarioState,
                                 TestStepResultValidator validator,
                                 TestStepResultConverter testStepResultConverter) {
        super(testScenarioState, validator, testStepResultConverter);
        this.tcpChannels = tcpChannels;
    }

    @Override
    protected Object processTestStep(TcpTestStep testStep) {
        Object response = null;
        Optional<String> sendTo = testStep.getSendTo();
        Optional<String> receiveFrom = testStep.getReceiveFrom();
        Optional<String> respondTo = testStep.getReceiveFrom();

        sendTo.ifPresent(channel -> send(testStep.getTestData(), channel));

        if (receiveFrom.isPresent()) {
            TcpChannel tcpChannel = tcpChannels.get(receiveFrom.get());
            response = runTask(tcpChannel::read, r -> r.isPresent() || tcpChannel.isClosed(), 5, 30).orElse(null);

            respondTo.ifPresent(s -> send(testStep.getTestData(), s));
        }
        return response;
    }

    private void send(Object testData, String channel) {
        TcpChannel tcpChannel = tcpChannels.get(channel);
        tcpChannel.send(testData);
    }

    @Override
    public Class<TcpTestStep> getTestStepClass() {
        return TcpTestStep.class;
    }

    @Override
    public void close() {
        tcpChannels.values().forEach(TcpChannel::close);
    }

    public static class Builder extends BaseTestStepProcessor.Builder<TcpTestStep> {

        private static final String TCP_CONFIG_KEYWORD = "tcp";

        private Map<String, TcpChannel> tcpChannels;

        public Builder() {
            tcpChannels = new HashMap<>();
        }

        public Builder withTcpChannel(TcpDetails tcpDetails) {
            this.tcpChannels.putIfAbsent(tcpDetails.getChannelId(), new TcpChannel(tcpDetails));
            return this;
        }

        public Builder withTcpChannels(Config config) {
            if (config.hasPath(TCP_CONFIG_KEYWORD)) {
                ((List<Map<String, String>>) config.getAnyRefList(TCP_CONFIG_KEYWORD))
                        .forEach(rawHttpService ->
                                withTcpChannel(new TcpDetails(rawHttpService.get("channelId"),
                                        rawHttpService.get("host"), Integer.parseInt(rawHttpService.get("port")))));
            }
            return this;
        }

        public TestStepProcessor<TcpTestStep> build() {
            return new TcpTestStepProcessor(tcpChannels, testScenarioState, validator, testStepResultConverter);
        }
    }

}

File: skellig-test-step-processing-tcp/src/main/kotlin/org/skellig/teststep/processor/tcp/TcpTestStepProcessor.kt
package org.skellig.teststep.processor.tcp

import org.skellig.teststep.processing.processor.BaseTestStepProcessor
import org.skellig.teststep.processing.processor.TestStepProcessor
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.util.debug
import org.skellig.teststep.processing.util.info
import org.skellig.teststep.processing.util.logger
import org.skellig.teststep.processor.tcp.model.TcpTestStep

/**
 * TcpTestStepProcessor is a class that processes TCP test steps.
 * It processes the test step this way:
 * 1) Sends a message ([TcpTestStep.testData]) to [TcpTestStep.sendTo]
 * 2) Reads a message ([TcpTestStep.testData]) from [TcpTestStep.readFrom]. If [TcpTestStep.respondTo] defined, then
 * responds with message ([TcpTestStep.testData]) to [TcpTestStep.respondTo] if the received message is valid
 * based on [TcpTestStep.validationDetails].
 *
 * All send, read and respond operations are performed in parallel per each channel but blocked until they are all finished.
 *
 * @param tcpChannels A map of TCP channels where the key is the channel ID and the value is the TCPChannel instance.
 * @param testScenarioState The state of the test scenario.
 */
open class TcpTestStepProcessor(
    private val tcpChannels: Map<String, TcpChannel>,
    testScenarioState: TestScenarioState?
) : BaseTestStepProcessor<TcpTestStep>(testScenarioState!!) {

    private val log = logger<TcpTestStepProcessor>()

    override fun processTestStep(testStep: TcpTestStep): Any? {
        var response: Any? = null
        val sendTo = testStep.sendTo
        val readFrom = testStep.readFrom
        val respondTo = testStep.respondTo

        sendTo?.let {
            log.info(testStep, "Start to send message of test step '${testStep.name}' to TCP channels $sendTo")
            send(testStep, it)
        }

        readFrom?.let {
            log.info(testStep, "Start to read message of test step '${testStep.name}' from TCP channels $readFrom")
            response = read(testStep, readFrom)

            respondTo?.let {
                if (isValid(testStep, response)) {
                    log.info(testStep, "Respond to received message to TCP channels '$respondTo'")
                    send(testStep, it)
                }
            }
        }
        return response
    }

    private fun read(testStep: TcpTestStep, channels: Set<String?>): Map<*, Any?> {
        val tasks = channels.associateWith {
            {
                val channel = tcpChannels[it] ?: error(getChannelNotExistErrorMessage(it))
                log.debug(testStep) { "Start to read message from TCP channel '$it'" }
                val message = channel.read(testStep.timeout, testStep.readBufferSize)
                log.debug(testStep) { "Received message from TCP channel '$it'" }
                message
            }
        }
        return runTasksAsyncAndWait(tasks, testStep)
    }

    private fun send(testStep: TcpTestStep, channels: Set<String>) {
        val tasks = channels.associateWith {
            {
                val channel = tcpChannels[it] ?: error(getChannelNotExistErrorMessage(it))
                log.debug(testStep) { "Send message to TCP channel '$it'" }
                channel.send(testStep.testData)
                log.debug(testStep) { "Message has sent to TCP channel '$it'" }
                "sent"
            }
        }
        runTasksAsyncAndWait(tasks, testStep)
    }

    override fun getTestStepClass(): Class<TcpTestStep> {
        return TcpTestStep::class.java
    }

    override fun close() {
        log.info("Close TCP Test Step Processor and all connections to channels")
        tcpChannels.values.forEach { it.close() }
    }

    private fun getChannelNotExistErrorMessage(channelId: String?) =
        "Channel '$channelId' was not registered in TCP Test Step Processor"

    class Builder : BaseTcpProcessorBuilder<TcpTestStep>() {
        override fun build(): TestStepProcessor<TcpTestStep> {
            return TcpTestStepProcessor(tcpChannels, testScenarioState)
        }
    }
}